pyanaconda/modules/payloads/source/url/url.py

```diff
@@ -0,0 +1,162 @@
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+from pyanaconda.core.signal import Signal
+from pyanaconda.core.util import ProxyString, ProxyStringError
+from pyanaconda.modules.common.errors import InvalidValueError
+from pyanaconda.modules.common.structures.payload import URLData, SSLConfigurationData
+from pyanaconda.modules.payloads.constants import SourceType, URLType
+from pyanaconda.modules.payloads.source.source_base import PayloadSourceBase
+from pyanaconda.modules.payloads.source.url.url_interface import URLSourceInterface
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class URLSourceModule(PayloadSourceBase):
+    """The URL source payload module."""
+
+    def __init__(self):
+        super().__init__()
+        self._url = URLData()
+        self.url_changed = Signal()
+
+        self._ssl_verification_enabled = True
+        self.ssl_verification_enabled_changed = Signal()
+
+        self._ssl_configuration = SSLConfigurationData()
+        self.ssl_configuration_changed = Signal()
+
+        self._proxy = ""
+        self.proxy_changed = Signal()
+
+    def is_ready(self):
+        """This source is ready for the installation to start."""
+        # FIXME: always true is correct but it will block change of payload source. Find solution!
+        return True
+
+    @property
+    def type(self):
+        """Get type of this source."""
+        return SourceType.URL
+
+    def for_publication(self):
+        """Get the interface used to publish this source."""
+        return URLSourceInterface(self)
+
+    def set_up_with_tasks(self):
+        """Set up the installation source.
+
+        :return: list of tasks required for the source setup
+        :rtype: [Task]
+        """
+        return []
+
+    def tear_down_with_tasks(self):
+        """Tear down the installation source.
+
+        :return: list of tasks required for the source clean-up
+        :rtype: [Task]
+        """
+        return []
+
+    @property
+    def url(self):
+        """Get URL data.
+
+        This data contains type of url link and the link.
+
+        :rtype: URLData data structure
+        """
+        return self._url
+
+    def set_url(self, url):
+        """Set URL data.
+
+        :param url: data structure containing url and type
+        :type url: URLData data structure
+        """
+        try:
+            URLType(url.type)
+        except ValueError:
+            raise InvalidValueError("Invalid source type set '{}'".format(url.type))
```

Comment:
Maybe, I would move this code to a  new method `_validate_url`.

Reply:
Good point. Fixed.